server/src/test/java/org/elasticsearch/cluster/routing/ShardRoutingTests.java

```diff
@@ -11,16 +11,178 @@
+            case 0 -> new ShardId(
+                randomValueOtherThan(instance.getIndexName(), ESTestCase::randomIdentifier),
+                instance.shardId().getIndex().getUUID(),
+                instance.id()
+            );
+            case 1 -> new ShardId(
+                instance.getIndexName(),
+                randomValueOtherThan(instance.shardId().getIndex().getUUID(), UUIDs::randomBase64UUID),
+                instance.id()
+            );
+            case 2 -> new ShardId(
+                instance.getIndexName(),
+                instance.shardId().getIndex().getUUID(),
+                randomValueOtherThan(instance.id(), () -> randomIntBetween(0, 99))
+            );
+            default -> throw new RuntimeException("unreachable");
+        };
+        return new ShardRouting(
+            newShardId,
+            instance.currentNodeId(),
+            instance.relocatingNodeId(),
+            instance.primary(),
+            instance.state(),
+            instance.recoverySource(),
+            instance.unassignedInfo(),
+            instance.relocationFailureInfo(),
+            instance.allocationId(),
+            instance.getExpectedShardSize()
+        );
+    }
+
+    private static ShardRouting mutateState(ShardRouting instance) {
+        var newState = randomValueOtherThan(instance.state(), () -> randomFrom(ShardRoutingState.values()));
+        return new ShardRouting(
+            instance.shardId(),
+            newState == ShardRoutingState.UNASSIGNED ? null : requireNonNullElseGet(instance.currentNodeId(), ESTestCase::randomIdentifier),
+            newState == ShardRoutingState.UNASSIGNED || newState == ShardRoutingState.STARTED
+                ? null
+                : requireNonNullElseGet(instance.relocatingNodeId(), ESTestCase::randomIdentifier),
+            instance.primary(),
+            newState,
+            newState != ShardRoutingState.UNASSIGNED && newState != ShardRoutingState.INITIALIZING
+                ? null
+                : requireNonNullElseGet(
+                    instance.recoverySource(),
+                    () -> TestShardRouting.buildRecoveryTarget(instance.primary(), newState)
+                ),
+            newState != ShardRoutingState.UNASSIGNED
+                ? instance.unassignedInfo()
+                : requireNonNullElseGet(instance.unassignedInfo(), () -> TestShardRouting.buildUnassignedInfo(newState)),
+            instance.relocationFailureInfo(),
+            switch (newState) {
+            case UNASSIGNED -> null;
+            case INITIALIZING, STARTED -> requireNonNullElseGet(instance.allocationId(), AllocationId::newInitializing);
+            case RELOCATING -> AllocationId.newRelocation(requireNonNullElseGet(instance.allocationId(), AllocationId::newInitializing));
+            },
+            instance.getExpectedShardSize()
+        );
+    }
+
+    private static ShardRouting mutateCurrentNodeId(ShardRouting instance) {
```

Comment:
Ideally we should only mutate the current node ID if the shard is assigned, and in that case we shouldn't mutate the state or anything else.

Reply:
I agree, however I do not think this is always possible with constraints in place. For example if current node is null (and shard is therefore unassigned) we need to mutate the state and all dependencies in order to set something

Reply:
Yep, so we should pick and choose the mutations to apply according to the state.

Reply:
I wish we have scala `case 0 if condition -> `